pkg/reconciler/choice/choice.go

```diff
@@ -0,0 +1,265 @@
+)
+
+const (
+	reconciled         = "Reconciled"
+	reconcileFailed    = "ReconcileFailed"
+	updateStatusFailed = "UpdateStatusFailed"
+)
+
+type Reconciler struct {
+	*reconciler.Base
+
+	// listers index properties about resources
+	choiceLister       listers.ChoiceLister
+	tracker            tracker.Interface
+	addressableTracker duck.AddressableTracker
+	subscriptionLister eventinglisters.SubscriptionLister
+}
+
+// Check that our Reconciler implements controller.Reconciler
+var _ controller.Reconciler = (*Reconciler)(nil)
+
+// Reconcile compares the actual state with the desired, and attempts to
+// reconcile the two. It then updates the Status block of the Choice resource
+// with the current Status of the resource.
+func (r *Reconciler) Reconcile(ctx context.Context, key string) error {
+	logging.FromContext(ctx).Debug("reconciling", zap.String("key", key))
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		logging.FromContext(ctx).Error("invalid resource key", zap.Error(err))
+		return nil
+	}
+
+	// Get the Choice resource with this namespace/name
+	original, err := r.choiceLister.Choices(namespace).Get(name)
+	if apierrs.IsNotFound(err) {
+		// The resource may no longer exist, in which case we stop processing.
+		logging.FromContext(ctx).Error("Choice key in work queue no longer exists")
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	// Don't modify the informers copy
+	choice := original.DeepCopy()
+
+	// Reconcile this copy of the Choice and then write back any status
+	// updates regardless of whether the reconcile error out.
+	reconcileErr := r.reconcile(ctx, choice)
+	if reconcileErr != nil {
+		logging.FromContext(ctx).Error("Error reconciling Choice", zap.Error(reconcileErr))
+		r.Recorder.Eventf(choice, corev1.EventTypeWarning, reconcileFailed, "Choice reconciliation failed: %v", reconcileErr)
+	} else {
+		logging.FromContext(ctx).Debug("Successfully reconciled Choice")
+		r.Recorder.Eventf(choice, corev1.EventTypeNormal, reconciled, "Choice reconciled")
+	}
+
+	if _, updateStatusErr := r.updateStatus(ctx, choice); updateStatusErr != nil {
+		logging.FromContext(ctx).Warn("Error updating Choice status", zap.Error(updateStatusErr))
+		r.Recorder.Eventf(choice, corev1.EventTypeWarning, updateStatusFailed, "Failed to update choice status: %s", key)
+		return updateStatusErr
+	}
+
+	// Requeue if the resource is not ready:
+	return reconcileErr
+}
+
+func (r *Reconciler) reconcile(ctx context.Context, p *v1alpha1.Choice) error {
+	p.Status.InitializeConditions()
+
+	// Reconciling choice is pretty straightforward, it does the following things:
+	// 1. Create a channel fronting the whole choice
+	// 2. For each of the Cases:
+	//     2.1 create a Subscription to the previous Channel and subscribe the filter
+	//     2.2 create a Subscription to the filter Channel and subcribe the subscriber
```

Comment:
can't step 4 be done here and set the reply as part of this step?

Reply:
... time passes... looks like there's no step 4 and it's instead done as part of the step 2.2 above?